src/arch/riscv/kernel/boot.c

```diff
@@ -234,6 +234,13 @@ static BOOT_CODE bool_t try_init_kernel(
     /* If a DTB was provided, pass the data on as extra bootinfo */
     p_region_t dtb_p_reg = P_REG_EMPTY;
     if (dtb_size > 0) {
+#ifdef CONFIG_PLAT_ROCKETCHIP
+        /* The softcore rocketchip instantiation doesn't work well when this
+         * page isn't reserved. Round up so the whole page is reserved to
+         * avoid the problem
+         */
+        dtb_size = ROUND_UP(dtb_size, PAGE_BITS);
```

Comment:
Can you be more specific, what is the problem then? For the user mode mapping, we will have to use full pages anyway. Maybe we should run up for all platforms?

Reply:
I didn't go super in-depth here to figure out _why_ this was happening, but I can tell you the behavior we were seeing.

We ran yanyan's VMM just fine with an older codebase. When rebasing onto this branch, we ran into a couple issues. One was the 8 byte padding, the other was that user-space generated an "virtual address already mapped" error. By reverting a few old commits, things would work properly, but that's not a good solution for larger commits in my opinion.

I was able to get past the virtual address error by increasing the static allocator mempool size, but I'd immediately get a RAM fault when trying to run the VM. 

So I looked at what was happening in the commits we needed to revert, and one of them removed the rounding for the device tree. Like you mentioned, userspace needs to map in full pages anyways, so if a portion of a page is reserved, the entire page should be. My best guess is that there was an overlap somewhere there shouldn't have been, but like I said, I didn't grab GDB and look at every memory map, I just found the solution.

Reply:
I've created https://github.com/seL4/seL4/issues/972 as a follow-up task to investigate this. So it's not necessarily a merge blocker to have this conditional code for now.

Reply:
Thanks, Axel!